### ISSUE TITLE
DAP server abruptly disconnecting

### ISSUE THREAD

**Issue by trevorwhitney** [go-delve/delve#3825]
1. What version of Delve are you using (`dlv version`)? 1.23.1
2. What version of Go are you using? (`go version`)? 1.23.2
3. What operating system and processor architecture are you using? Mac OS 15.0.1
4. What did you do?
  I tried to run a launch config for a `main.go` file with args, from neovim, using `nvim-dap` and `nvim-dap-go`
6. What did you expect to see?
  I expected to see the program launch, as it has for years.
8. What did you see instead?
  The Dap server quit abruptly, [nvim-dap logs](https://gist.github.com/trevorwhitney/92c87013a9f83149506fec03d60a2b86) and [dap logs](https://gist.github.com/trevorwhitney/152c616365fa3df34481ca6ccb54c713)

I don't understand why the DAP server is terminating, or how to get more information to troubleshoot this?

**Comment by aarzilli** [comment]
It doesn't look like any breakpoints are being set and the program just
runs and terminates.

Il mar 8 ott 2024, 23:03 Trevor Whitney ***@***.***> ha
scritto:

**Comment by trevorwhitney** [comment]
It's a server so it should just keep running until killed. I've tried setting breakpoints, and I get the same output, with the same result of the dap server just terminating. Any ideas on other ways to get more verbose output to troubleshoot this?

**Comment by trevorwhitney** [comment]
Thanks for helping out on this, it's a huge hit to my workflow to not have a working debugger.

Debugging from the command line (not using DAP) seems to work fine:

```
CGO_FLAGS="-g -O0" dlv --log --log-output debugger --build-flags "-gcflags='all=-N -l'" debug
```

This produced a binary I was able to run with the same args and it came up fine. I've only ever used DAP with an editor. I tested out debugging in VSCode (which I understand uses delve's DAP protocol as well), and that worked. Is there an easy way to test the DAP functionality from the command line that you know of?

**Comment by trevorwhitney** [comment]
I tried running the debugger on a trivial program and it worked

```
package main

import (
	"fmt"
	"os"
)

func main() {
	name := os.Args[1]
	fmt.Printf("hello %s\n", name)
}
```
Up to this line, everything is the same. The working program prints `[debug-adapter stderr] 2024-10-09T08:51:41-06:00 debug layer=debugger callInjection protocol on` when the non-working program prints `[debug-adapter stderr] 2024-10-09T08:53:43-06:00 debug layer=dap [-> to client]{"seq":0,"type":"event","event":"terminated","body":{}}`. I'm assuming this is some `delve` error? Is there any way to figure out why the `callInjection` protocol is failing?

**Comment by aarzilli** [comment]
>  I'm assuming this is some delve error? Is there any way to figure out why the callInjection protocol is failing?

That's irrelevant, it just prints that when it stops at a breakpoint. 

> This produced a binary I was able to run with the same args and it came up fine

You don't need to manually specify CGO_CFLAGS and -gcflags when calling dlv. What did you do exactly with the debugger called from the command line?

**Comment by trevorwhitney** [comment]
@aarzilli thanks again for all your help on this, I think I finally got to the root of it and it doesn't seem to be related to `delve`.

For anyone else who finds themselves here. The project I was failing to run in the debugger was `grafana/loki`. Once I got the simple program running, I tried to figure out if there was a commit that broke this for me in `grafana/loki` and landed on [this one](https://github.com/grafana/loki/commit/c8e6a9d38f36ccf1f32e634765bb2363628f3710).

https://github.com/grafana/loki/pull/14437 reverts `shirou/gopsutil` back to `v4.24.8`, which is before they merged https://github.com/shirou/gopsutil/pull/1702 which I'm suspecting was the problem. I filed a terribly hard to reproduce issue [here](https://github.com/shirou/gopsutil/issues/1734), but I don't think I understand the problem any better than this to be able to provide more detail.
